boat_simulator/common/sensors.py

```diff
@@ -60,98 +81,179 @@ def read(self, key: str) -> Any:
+        enable_delay (bool): Enables delay for fields. False by default.
     """
 
     wind: ScalarOrArray
-    wind_noisemaker: WindSensorGenerators = None
+
+    def __init__(
+        self,
+        wind: ScalarOrArray,
+        stdev: List[Scalar] = [1.0, 1.0],
+        enable_noise: bool = False,
+        enable_delay: bool = False,
+    ) -> None:
+        super().__init__(enable_noise=enable_noise, enable_delay=enable_delay)
+        self._wind = wind
+
+        # TODO: Refactor the initialization of data fields and their respective delay controls.
+        # Warning: this is not easy!
+
+        self.wind_queue_next: bool = False
+        self.wind_next_value: ScalarOrArray = wind
+        self.wind_noisemaker: MVGaussianGenerator = MVGaussianGenerator(
+            mean=np.array([0, 0]), cov=np.diag(np.power(stdev, 2))
+        )
 
     @property  # type: ignore
     def wind(self) -> ScalarOrArray:
         # TODO: Ensure attribute value and noisemakers are using the same value shape.
         # - wind scalars should add with noise scalars.
         # - wind vectors should add with noise vectors.
-        # Could consider using a __post_init__ function for this
+
         return (
             self._wind + self.wind_noisemaker.next()  # type: ignore
-            if self.wind_noisemaker is not None
+            if self.enable_noise
             else self._wind
         )
 
     @wind.setter
     def wind(self, wind: ScalarOrArray):
-        self._wind = wind
+
+        if not self.enable_delay:
+            self._wind = wind
+            return
+
+        if self.wind_queue_next:
+            self._wind = self.wind_next_value
+        else:
+            self.wind_queue_next = True
+
+        self.wind_next_value = wind
 
 
-@dataclass
 class GPS(Sensor):
     """
     Abstraction for GPS.
 
-    # TODO: Add delay functions.
-
     Properties:
         lat_lon (NDArray): Boat latitude and longitude (2x1 array)
         speed (Scalar): Boat speed
         heading (Scalar): Boat heading
-        lat_lon_noisemaker (Optional[GaussianGenerator | ConstantGenerator]):
-        Noise function to emulate sensor noise in latitude and longitude readings
-        speed_noisemaker (Optional[GaussianGenerator | ConstantGenerator]):
-        Noise function to emulate sensor noise in speed readings
-        heading_noisemaker (Optional[GaussianGenerator | ConstantGenerator]):
-        Noise function to emulate sensor noise in heading readings
+        enable_noise (bool): Enables noise for fields. False by default.
+        enable_delay (bool): Enables delay for fields. False by default.
     """
 
     lat_lon: NDArray
     speed: Scalar
     heading: Scalar
 
-    lat_lon_noisemaker: GPSGenerators = None
-    speed_noisemaker: GPSGenerators = None
-    heading_noisemaker: GPSGenerators = None
+    def __init__(
+        self,
+        lat_lon: NDArray,
+        speed: Scalar,
+        heading: Scalar,
+        stdev: Scalar = 1,
```

Comment:
What happens when it's zero? Similar question for the wind sensor

Reply:
behaves like the constant generator.